manager/src/main/java/io/gingersnapproject/Caches.java

```diff
@@ -60,6 +65,28 @@ private void actualRemove(String name, String key, Uni<String> value) {
       }
    }
 
+   public  Uni<String> denormalizedPut(String name, String key, String value) {
+      Table table = databaseHandler.table(configuration.rules().get(name).connector().table());
+      if (!table.foreignKeys().isEmpty()) {
+         // We have foreign keys to resolve
+         Json json = Json.read(value);
+         UniJoin.Builder<Object> builder = Uni.join().builder();
+         for (ForeignKey fk : table.foreignKeys()) {
+            // TODO: handle composite fks
+            Json fkJson = json.atDel(fk.columns().get(0));
+            if (fkJson != null) {
+               String fkId = fkJson.asString();
+               String fkRule = databaseHandler.tableToRuleName(fk.refTable());
+               builder.add(get(fkRule, fkId).map(Json::read).map(j -> json.set(fkRule, j)));
```

Comment:
Should we use the `refTable` here instead of `fkRule` in the `json.set`?

Reply:
This line is also not thread safe. It is possible that we have multiple threads returning from the Uni and writing to the json variable concurrently.

Reply:
We probably want to copy the `Json` class over from Infinispan and make it thread-safe